### ISSUE TITLE
GEMM-ceiling probe hardcodes bfloat16, so the throughput forecast is pessimistic on every pre-Ampere card

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#617]
## Summary

`measure_gemm_tflops`'s probe allocates its matrices as `bfloat16` unconditionally, regardless of the dtype the run resolved. On a pre-Ampere card (compute capability < 8.0 — T4, P100, V100, GTX 16xx, Quadro T-series) there are no native bf16 tensor cores, so the probe measures **emulated** bf16 throughput while training will actually run in fp16. The GEMM ceiling, and therefore the throughput bracket printed in the layer-streaming pre-flight panel, is pessimistic on that entire GPU generation.

Reported by @ArtisanalAutomationsFactories as a secondary finding while root-causing #558 (which turned out to be a duplicate of #425 and unrelated to this). Confirmed.

```python
# src/soup_cli/utils/layer_stream_runtime.py:1289-1290
left  = torch.randn(size, size, device=device, dtype=torch.bfloat16)
right = torch.randn(size, size, device=device, dtype=torch.bfloat16)
```

Observed forecast on a Quadro T2000 (cc 7.5): **26-39 tok/s**, which the reporter expects to be an underestimate for cards in this class once the probe uses the dtype the run will actually use.

## Severity: cosmetic, and the boundary is worth stating

This is display-only. `measure_gemm_tflops` feeds `GemmCeiling` -> `forecast_stream_throughput`, which prints a bracket in the pre-flight panel. It is not on the training path, it cannot reach the optimizer, and it does not affect correctness, VRAM budgeting, or the fit refusal.

It still matters: the forecast is one of the few numbers an operator sees *before* committing to a run, and one that is systematically wrong on a whole GPU generation trains people to ignore the panel. This project has an issue open (#405) about a metric that cannot move; a metric that moves in the wrong direction on pre-Ampere is the same category of problem.

## Proposed fix

```python
dtype = _torch_dtype(resolve_stream_dtype(str(device)))
left  = torch.randn(size, size, device=device, dtype=dtype)
right = torch.randn(size, size, device=device, dtype=dtype)
```

`resolve_stream_dtype` already returns `"float16"` on a card whose `cuda_supports_bf16()` is false (the #385/#387 work, where the recorded trap is that `torch.cuda.is_bf16_supported()` defaults to `including_emulation=True` and answers **True** on a T4 — use the existing helper, do not re-derive this).

Note `GemmCeiling`'s docstring says "a dense **bf16** GEMM rate measured on THIS card". If the dtype becomes variable, the docstring and any prose that quotes it must say which dtype the number is in, or the ceiling stops being comparable between two runs on different cards.

## Acceptance criteria

- [ ] The probe allocates in the dtype the run resolved, via `resolve_stream_dtype`, not a literal
- [ ] `GemmCeiling` records **which** dtype it measured, and the panel says so — a bracket whose unit is unstated is worse than one that is consistently wrong
- [ ] A test that fails when the literal is restored: force `resolve_stream_dtype` to `"float16"` and assert the allocated dtype follows. This must not require a GPU — the allocation call is mockable
- [ ] A control asserting the bf16 path still allocates bf16 on a card that supports it, so a fix that hardcodes `float16` instead is not an improvement
- [ ] No change to the VRAM fit decision or to any published number in `benchmarks/` — this is display-only, and if a gate number moves, something else is wrong

## Notes

@ArtisanalAutomationsFactories, this is yours if you want it — you found it, and your proposed shape is what is written above. Say so on the issue and it is claimed; if you would rather not, it is a good first issue for someone else.

**What this issue does NOT cover:** whether the forecast is *accurate* on pre-Ampere once the dtype is right. Nobody here has a Turing card to measure it on, so the acceptance criteria are about the probe using the correct dtype, not about the resulting number being validated. Re-measuring the bracket on pre-Ampere hardware is a separate question and belongs with #325 / #395.

**Comment by Samearth17** [comment]
I would like to take #617 if it’s still available.

I will keep the scope to the proposed display-only fix: resolve the GEMM probe dtype through `resolve_stream_dtype`, carry that dtype in `GemmCeiling`, update the forecast panel to state the measured dtype, and add CPU/mock coverage for both the pre-Ampere fp16 path and the native bf16 control.

I will leave the published benchmark numbers and VRAM-fit logic unchanged.

**Comment by codewithchotu** [comment]
Hi! I noticed #617 is currently not formally assigned, but I saw that @Samearth17 has opened PR #648 for it. Is the issue still available for another contributor, or should I look for another issue?

**Comment by MakazhanAlpamys** [comment]
@codewithchotu — thanks for checking the queue before starting rather than after; that is exactly the right order and it is why this repo has almost no duplicated work.

**#617 is taken.** @Samearth17 claimed it on 2026-08-31 with a scoped plan and opened #648, which is in review right now (I left four items on it this morning). I have assigned the issue to them so the state is visible from the issue page and nobody else has to ask.

**Free work that fits a first PR here**, in the order I would pick:

1. **#275 — a task-variant recipe.** The cheapest genuinely useful first PR in this repo: one `RecipeMeta(...)` entry in `src/soup_cli/recipes/catalog.py`, one parametrized test row, the catalog count bumped in the three places that assert it. The acceptance criteria are written out on the issue. Five 2026 families currently ship **SFT only**, so a DPO / GRPO / pretrain variant for any of them is unclaimed: `minimax-m3`, `mistral-large-3`, `kimi-k2.5`, `qwen3.6-27b`, `deepseek-v4-pro`. Note @Srinivasan8888 has `deepseek-v4-flash-dpo` in flight, so avoid that one.
2. **#659** — `soup doctor` never reports MLX. Small, self-contained, and the diagnosis is already done in the issue. @Srinivasan8888 found it while working on something else and has not claimed it, so I am offering it here **subject to their right of first refusal** — I will confirm on that issue before you invest time.

**How to claim, whichever you pick:** comment on the issue naming the specific thing you are taking (the recipe name, not just "#275"), and do the overlap check first — `gh pr list --search <keyword> --state all` catches claims that only ever appeared in a PR title. That habit is the reason this project has a low collision rate, and it is the single thing I would ask a new contributor to copy.

Worth reading before your first PR: `CONTRIBUTING.md`, and the merge comment on any recent PR — the review bar here is a mutation test of your own fix, not just a green CI.

**Comment by MakazhanAlpamys** [comment]
Fixed in `f14bf42` by @Samearth17 in #648. All five acceptance criteria met, each verified by execution rather than by reading the diff:

- [x] **Probe allocates via `resolve_stream_dtype`, not a literal.** Restoring the hard-coded `torch.bfloat16` fails `test_gemm_uses_resolved_stream_dtype[float16]`.
- [x] **`GemmCeiling` records the dtype and the panel says so.** Deleting `using {ceiling.dtype}` from the panel string fails both new `TestIssue617PanelDtype` cases — and fails them on a CPU-only run, so all nine CI cells catch it.
- [x] **A test that fails when the literal is restored, without needing a GPU.** Both mutations above die under `CUDA_VISIBLE_DEVICES=-1`.
- [x] **A bf16 control**, so a fix that hard-codes `float16` instead is not mistaken for an improvement — parametrized `[float16, bfloat16]`.
- [x] **No change to the fit decision and no published number moved.** `ceiling` is consumed only inside the display-only branch; `decide_stream_fit` runs earlier on predicted/available and is untouched. `benchmarks/` is unmodified.

Try:

```bash
soup train --config soup.yaml   # with training.stream_layers: true
```

The forecast line now names the dtype it measured:

```
forecast  5686-8362 tok/s — a compute-bound bound, not a promise
          (from 6.75 TFLOPS measured on this card now using bfloat16 @ 862 MHz)
```

**The trap that let this survive is worth recording**: `torch.cuda.is_bf16_supported()` returns True on a T4 through emulation, so the obvious verification agrees with the wrong answer. Routing the probe through the same `resolve_stream_dtype` the streaming path itself uses means the two can no longer disagree.

Nobody here has a Turing card, so @Samearth17 drove the pre-Ampere direction by faking exactly one thing — `is_bf16_supported(including_emulation=False)` returning False, which is a T4's truth — and showed the real `soup train` resolving `float16`, allocating fp16, and printing it. One fake, at the hardware boundary, everything downstream real.

## Two follow-ups this leaves open

The issue already scoped out whether the forecast is *accurate* on pre-Ampere once the dtype is right, and that stands. But there is one thing the issue got wrong, discovered while reviewing the fix: it says nobody has a Turing card, and in fact `benchmarks/run-t4-colab-free-tier.md:64` carries a published T4 record — `forecast 31-46 tok/s (from 2.20 TFLOPS measured on this card @ 1185 MHz)` — measured under the emulated-bf16 probe this change removes. So there **is** one published pre-Ampere number, and this fix makes it stale in substance rather than merely in wording. It is a historical transcript and correctly left untouched by #648; I will file the re-measurement separately.

Second, a limitation of the tests rather than of the fix: the torch stub sets `torch.bfloat16 = "bfloat16"`, so the assertions cannot distinguish a real `torch.dtype` lookup from the raw name being passed through. A mutation doing exactly that is caught with real CUDA and invisible on every CI cell. That is inherent to CI having no GPU runner, and it is recorded in the merge commit.

**Comment by MakazhanAlpamys** [comment]
Filed the T4 re-measurement as #714, as promised above. It is narrower than #325 / #395: one published record, one card, one number — and unlike most pre-Ampere work it is actually reachable, since the original measurement was taken on Colab's free tier.
